sycl/source/detail/queue_impl.cpp

```diff
@@ -192,11 +192,18 @@ void queue_impl::wait(const detail::code_location &CodeLoc) {
   TelemetryEvent = instrumentationProlog(CodeLoc, Name, StreamID, IId);
 #endif
 
-  std::lock_guard<mutex_class> Guard(MMutex);
-  for (std::weak_ptr<event_impl> &EventImplWeakPtr : MEvents) {
-    if (std::shared_ptr<event_impl> EventImplPtr = EventImplWeakPtr.lock())
-      EventImplPtr->wait(EventImplPtr);
+  std::vector<std::shared_ptr<event_impl>> Events;
+  {
+    std::lock_guard<mutex_class> Guard(MMutex);
+    for (std::weak_ptr<event_impl> &EventImplWeakPtr : MEvents)
+      if (std::shared_ptr<event_impl> EventImplPtr = EventImplWeakPtr.lock())
+        Events.push_back(EventImplPtr);
+  }
+
+  for (std::shared_ptr<event_impl> &Event : Events) {
+    Event->wait(Event);
   }
+
```

Comment:
What's achieved by this change?
Is it required for the fix of the issue? If not, I suggest moving it to a separate patch.

Reply:
The change fixes the situation which leads to deadlock which I encountered:
1) queue::wait call
2) queue's mutex MMutex is locked 
3) host task throws an exception which should be reported as asynchronous.
4) reportAsyncException is called
5) thread encounters MMutex in this function which is locked
6) ???
7) _Deadlock!_

Reply:
I believe this deadlock should also be represented in description of this patch.

Reply:
Is it sporadic deadlock or stable one?
Either way, we ought to have a test which fails without your patch and succeeds with your patch included.

Reply:
It was a stable deadlock. The patch includes a test which hangs without this change.

Reply:
Is it `test6` in `sycl/test/host-interop-task/interop-task.cpp`? I believe this test is only for host-task's exception reproting. Please add a comment describing the test.

Reply:
Comment has been added.